execution/executor/src/db_bootstrapper.rs

```diff
@@ -32,14 +32,55 @@ pub fn bootstrap_db_if_empty<V: VMExecutor>(
         return Ok(None);
     }
 
-    Ok(Some(bootstrap_db::<V>(db, tree_state, genesis_txn)?))
+    let committer = calculate_genesis::<V>(db, tree_state, genesis_txn)?;
+    let waypoint = committer.waypoint;
+    committer.commit()?;
+    Ok(Some(waypoint))
 }
 
-pub fn bootstrap_db<V: VMExecutor>(
+pub struct GenesisCommitter<V: VMExecutor> {
+    executor: Executor<V>,
+    ledger_info_with_sigs: LedgerInfoWithSignatures,
+    waypoint: Waypoint,
+}
+
+impl<V: VMExecutor> GenesisCommitter<V> {
+    pub fn new(
+        executor: Executor<V>,
+        ledger_info_with_sigs: LedgerInfoWithSignatures,
+    ) -> Result<Self> {
+        let waypoint = Waypoint::new(ledger_info_with_sigs.ledger_info())?;
+
+        Ok(Self {
+            executor,
+            ledger_info_with_sigs,
+            waypoint,
+        })
+    }
+
+    pub fn ledger_info_with_sigs(&self) -> &LedgerInfoWithSignatures {
+        &self.ledger_info_with_sigs
+    }
+
+    pub fn waypoint(&self) -> Waypoint {
+        self.waypoint
+    }
+
+    pub fn commit(mut self) -> Result<()> {
+        self.executor
+            .commit_blocks(vec![genesis_block_id()], self.ledger_info_with_sigs)?;
+        info!("Genesis commited.");
+        // DB bootstrapped, avoid anything that could fail after this.
+
+        Ok(())
+    }
+}
+
+pub fn calculate_genesis<V: VMExecutor>(
     db: &DbReaderWriter,
     tree_state: TreeState,
     genesis_txn: &Transaction,
-) -> Result<Waypoint> {
+) -> Result<GenesisCommitter<V>> {
```

Comment:
This doesn't feel completely right. Maybe just the naming...

Reply:
will PR++ when I come up with better naming..